core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -431,6 +458,155 @@ object SparkSubmit {
+      ivySettings.addMatcher(new GlobPatternMatcher)
+
+      // the biblio resolver resolves POM declared dependencies
+      val br: IBiblioResolver = new IBiblioResolver
+      br.setM2compatible(true)
+      br.setUsepoms(true)
+      br.setName("central")
+
+      // We need a chain resolver if we want to check multiple repositories
+      val cr = new ChainResolver
+      cr.setName("list")
+      cr.add(br)
+
+      // Add an exclusion rule for Spark
+      val sparkArtifacts = new ArtifactId(new ModuleId("org.apache.spark", "*"), "*", "*", "*")
+      val sparkDependencyExcludeRule =
+        new DefaultExcludeRule(sparkArtifacts, ivySettings.getMatcher("glob"), null)
+      sparkDependencyExcludeRule.addConfiguration(conf)
+
+      // add any other remote repositories other than maven central
+      if (remoteRepos != null && remoteRepos.trim.nonEmpty) {
+        var i = 1
+        remoteRepos.split(",").foreach { repo =>
+          val brr: IBiblioResolver = new IBiblioResolver
+          brr.setM2compatible(true)
+          brr.setUsepoms(true)
+          brr.setRoot(repo)
+          brr.setName(s"repo-$i")
+          cr.add(brr)
+          logInfo(s"$repo added as a remote repository with the name: ${brr.getName}")
+          i += 1
+        }
+      }
+      ivySettings.addResolver(cr)
+      ivySettings.setDefaultResolver(cr.getName)
+      val ivy = Ivy.newInstance(ivySettings)
+      // Set resolve options to download transitive dependencies as well
+      val resolveOptions = new ResolveOptions
+      resolveOptions.setTransitive(true)
+      val retrieveOptions = new RetrieveOptions
+      // Turn downloading and logging off for testing
+      if (isTest) {
+        resolveOptions.setDownload(false)
+        resolveOptions.setLog(LogOptions.LOG_QUIET)
+        retrieveOptions.setLog(LogOptions.LOG_QUIET)
+      } else {
+        resolveOptions.setDownload(true)
+      }
+
+      // A Module descriptor must be specified. Entries are dummy strings
+      val md = DefaultModuleDescriptor.newDefaultInstance(
+        ModuleRevisionId.newInstance("org.apache.spark", "spark-submit-envelope", "1.0"))
+      md.setDefaultConf(conf)
+
+      md.addExcludeRule(sparkDependencyExcludeRule)
+
+      artifacts.foreach { mvn =>
+        val ri = ModuleRevisionId.newInstance(mvn.groupId, mvn.artifactId, mvn.version)
+        val dd = new DefaultDependencyDescriptor(ri, false, false)
+        dd.addDependencyConfiguration(conf, conf)
+        logInfo(s"${dd.getDependencyId} added as a dependency")
+        md.addDependency(dd)
+      }
+
+      // resolve dependencies
+      val rr: ResolveReport = ivy.resolve(md, resolveOptions)
+      if (rr.hasError) {
+        throw new RuntimeException(rr.getAllProblemMessages.toString)
+      }
+      // Log the callers for each dependency
+      rr.getDependencies.toArray.foreach { case dependency: IvyNode =>
+        logInfo(s"$dependency will be retrieved as a dependency for:")
```

Comment:
This `logInfo` call and the following call might have other logging statements interleaved, so I think it would be preferable to build up a single multi-line log string rather than making multiple calls here.